drivers/hmis/app/graphql/types/hmis_schema/assessment.rb

```diff
@@ -58,44 +59,51 @@ def role
 
     # EXPENSIVE! Do not use in batch
     def definition
-      project = load_ar_association(object, :project)
-
-      form_processor = load_ar_association(object, :form_processor)
-      # If this occurs, it may be an issue with MigrateAssessmentsJob, SaveAssessment, or SubmitAssessment
-      raise "Assessment without form processor: #{object.id}" unless form_processor.present?
-
-      # If definition is stored on form processor, return that.
-      # TODO: check if form is retired? For non-WIP non-custom assessments, we should
-      # really be choosing the "latest" form, which may not be the one on the FormProcessor.
+      # If definition is stored on Form Processor, return that. This is the Definition that was most recently used to submit the Assessment.
       definition = load_ar_association(form_processor, :definition)
-      # If there was no definition specified, which would occur if this is a migrated assessment, choose an appropriate one.
+      # If there was no definition on the Form Processor, which would occur if this is a migrated HUD Assessment, then choose an appropriate one:
       definition ||= Hmis::Form::Definition.find_definition_for_role(role, project: project)
+      # Apply filter context to filter out items that are not relevant for this project (HUD and Custom Rules)
       definition.filter_context = { project: project, active_date: object.assessment_date }
       definition
     end
 
+    # EXPENSIVE! Do not use in batch
+    def upgraded_definition_for_editing
+      return unless form_processor.definition_id # tiny optimization: avoid calling 'definition' if it will invoke find_definition_for_role twice
+
+      previous_definition = definition
+      # If original form is not retired, then we should stil use it for editing.
+      return unless previous_definition.retired?
+
+      # Find the published version of the previous definition. If this resolves nil, then the original form will be used.
+      published_definition = previous_definition.published_version
+      published_definition&.filter_context = { project: project, active_date: object.assessment_date }
+      published_definition
+    end
```

Comment:
This is the meaningful change. Very open to suggestions on the naming of field `assessment.upgradedDefinitionForEditing { ... }`.

Reply:
Okay, thanks for the ideas. I will leave it as-is!